In [44]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


### Get the number of unique streams per organizations from the stream_details csv, which captures organization, stream, card mapping

In [45]:
stream_details_df = pd.read_csv("../data/stream_details.csv", encoding = "ISO-8859-1")
print(stream_details_df.head())

   ORGANIZATION  STREAMID                STREAMNAME  CARDID  \
0            14       163  Lubes Concepts for Japan    1090   
1            14       163  Lubes Concepts for Japan    1091   
2            14       163  Lubes Concepts for Japan    1092   
3            14       163  Lubes Concepts for Japan    1093   
4            14       163  Lubes Concepts for Japan    1094   

                 CARDNAME  
0            ????????????  
1    ???????????? - ?????  
2    ???????????? - ?????  
3  ???????????? - ???????  
4    ???????????? - ?????  


### Get the number of unique organizations

In [46]:
print(stream_details_df["ORGANIZATION"].nunique())

2


### Get the number of unique streams

In [47]:
print(stream_details_df["STREAMID"].nunique())

840


### Get the number of cards per stream

In [48]:
cards_per_stream = stream_details_df.groupby('STREAMID')['CARDID'].nunique()
print(cards_per_stream)

STREAMID
163     15
167     14
171      6
172      4
173      8
186      6
199     12
201     10
202      1
203     15
204      1
205      8
206      9
218     21
219      1
228      4
229      1
230      1
231      8
232      8
233      6
236      1
237      1
244     18
245      7
246      8
247      1
248      3
315      3
316      1
        ..
1657    14
1659     2
1660    27
1661     1
1662     6
1663     7
1664     7
1665     4
1666     3
1668     2
1669     5
1670     8
1671     7
1672     3
1673     2
1674    17
1677     1
1678     1
1680     7
1682     1
1683     7
1684     2
1685     1
1686     1
1687     1
1688     1
1689     1
1690     1
1691     1
1694     2
Name: CARDID, Length: 840, dtype: int64


### Get the distribution of streams for each organization - group by the organization and then the number of streams

In [49]:
stream_details_by_org = stream_details_df.groupby('ORGANIZATION')['STREAMID'].nunique()
print(stream_details_by_org)

ORGANIZATION
14    120
28    720
Name: STREAMID, dtype: int64


In [50]:
users_details_df = pd.read_csv("../data/UserProfile_Customer1 and 2.csv", encoding = "ISO-8859-1")
print(users_details_df.head())

   org_id  User_Id  Gender  Dob     Location    City Country  \
0      14      245     NaN  NaN          NaN     NaN      SG   
1      14      337     NaN  NaN          NaN     NaN      SG   
2      14      267     NaN  NaN  Philippines  Manila      PH   
3      14      343     NaN  NaN          NaN     NaN      SG   
4      14      574     NaN  NaN          NaN     NaN      PH   

               Role user_timezone  
0  No Role Assigned     Singapore  
1  No Role Assigned     Singapore  
2  No Role Assigned     Singapore  
3  No Role Assigned     Singapore  
4  No Role Assigned     Singapore  


### Get the number of unique organizations from the user table

In [51]:
print(users_details_df["org_id"].nunique())

2


### Get the number of unique users from the user table

In [52]:
print(users_details_df["User_Id"].nunique())

22412


### Get the distribution of users for each organization - group by the organization and then the number of users

In [53]:
user_details_by_org = users_details_df.groupby('org_id')['User_Id'].nunique()
print(user_details_by_org)

org_id
14      242
28    22170
Name: User_Id, dtype: int64


#### Seems there is substantial representation from organization with org_id = 28

### Get the views per user and group by organization

In [54]:
content_views_df = pd.read_csv("../data/stream_views/0907to0908.csv", encoding = "ISO-8859-1")
print(content_views_df.head())

   USERID  CARDID      CARDTITLE  STREAMID STREAMNAME            TIMESTAMP  \
0    1019    3021  Best in Class       579  About HMD  2017-07-09 10:23:36   
1    1054    3021  Best in Class       579  About HMD  2017-07-09 14:41:25   
2     977    3021  Best in Class       579  About HMD  2017-07-09 13:35:38   
3    1337    3021  Best in Class       579  About HMD  2017-07-10 05:53:05   
4    1346    3021  Best in Class       579  About HMD  2017-07-10 11:23:59   

   MODULEID MODULENAME  ORGANIZATION  
0       473        HMD            28  
1       473        HMD            28  
2       473        HMD            28  
3       473        HMD            28  
4       473        HMD            28  


In [55]:
content_views_per_organization = content_views_df.groupby('ORGANIZATION')["STREAMID"].nunique()
print(content_views_per_organization)

ORGANIZATION
28    136
Name: STREAMID, dtype: int64


### So this file only has streams of organization 28
#### The number of unique streams that has been viewed
#### There are *136* streams in total in the file

In [56]:
unique_num_content_views_by_user = content_views_df.groupby('USERID')["STREAMID"].nunique()
print(unique_num_content_views_by_user)

USERID
950      3
958      4
959     10
961      6
963     21
964      2
977     22
1019     5
1020     1
1021     4
1027    10
1034     4
1035     1
1038     1
1039     2
1044     2
1046     5
1054    22
1058     4
1059    17
1063    20
1064    20
1065    13
1066     1
1067     3
1076     4
1078     5
1079     2
1080     5
1081    10
        ..
2760     2
2761    13
2762     9
2763    19
2764    11
2765     4
2767     5
2768     5
2769     4
2770     7
2771     5
2772     2
2773     8
2774     7
2775     1
2776     2
2777     2
2778     6
2782    13
2783     6
2787     2
2788     1
2789     3
2790     2
2792     5
2794    10
2797     9
2799    19
2800     2
2803     6
Name: STREAMID, Length: 1320, dtype: int64


In [57]:
# Generate a pivot table from the views data
content_views_by_user = pd.pivot_table(content_views_df, index = "USERID", columns = "STREAMID", values = "TIMESTAMP", aggfunc="count")

# fill the NA values with 0 - since there are no views for these contents
content_views_by_user = content_views_by_user.fillna(value = 0)

content_views_by_user.head()

STREAMID  579   580   581   582   583   584   585    586   587   588 ...   \
USERID                                                               ...    
950       0.0   0.0   0.0   0.0  12.0   0.0   0.0    0.0   0.0   0.0 ...    
958       0.0   5.0   0.0   0.0   0.0   0.0   0.0    0.0   0.0  14.0 ...    
959       0.0  31.0  24.0   7.0   0.0   6.0   0.0   32.0   0.0   0.0 ...    
961       0.0  14.0   9.0   0.0   0.0  16.0   0.0   34.0   0.0   0.0 ...    
963       7.0  36.0  21.0  24.0   0.0  28.0  21.0  101.0  51.0  44.0 ...    

STREAMID  786  787  788  789  790  793  795  796  797  798  
USERID                                                      
950       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
958       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
959       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
961       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
963       0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 136 columns]

In [58]:
content_views_by_user.to_csv("../data/generated/content_views_by_user_op.csv")

# the 5 quantiles
quantile_values = [0, 0.25, 0.5, 0.75, 1]

# quantiles by users
quantiles_by_users = content_views_by_user.quantile(quantile_values, axis = 0)

# quantiles by streams
quantiles_by_streams = content_views_by_user.quantile(quantile_values, axis = 1)

print(quantiles_by_users)
print(quantiles_by_streams)

STREAMID    579    580    581    582   583    584    585     586     587  \
0.00        0.0    0.0    0.0    0.0   0.0    0.0    0.0     0.0     0.0   
0.25        0.0    0.0    0.0    0.0   0.0    0.0    0.0     0.0     0.0   
0.50        0.0    4.0    2.0    2.0   0.0    1.0    0.0     0.0     0.0   
0.75        3.0   10.0    8.0    7.0   0.0    9.0    6.0    17.0    14.0   
1.00      222.0  433.0  491.0  437.0  12.0  739.0  648.0  7247.0  2222.0   

STREAMID     588 ...    786  787  788  789  790  793  795  796  797  798  
0.00         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.25         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.50         0.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
0.75        13.0 ...    0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1.00      1707.0 ...   14.0  6.0  6.0  6.0  6.0  8.0  1.0  4.0  2.0  1.0  

[5 rows x 136 columns]
USERID  950   958   959   961    963   964   977   1019  1020  1021  

### Considering the number of streams that each user has viewed

In [59]:
quantiles_by_num_views = unique_num_content_views_by_user.quantile(quantile_values)
print(quantiles_by_num_views)

0.00     1.0
0.25     3.0
0.50     7.0
0.75    16.0
1.00    41.0
Name: STREAMID, dtype: float64


In [60]:
for streamId in content_views_by_user:
    
    if cards_per_stream.get(streamId):
        content_views_by_user[streamId] = content_views_by_user[streamId] / cards_per_stream[streamId]
    else:
        print("No cards for the stream: " + str(streamId))
        
content_views_by_user.head()

No cards for the stream: 583
No cards for the stream: 601
No cards for the stream: 605
No cards for the stream: 621
No cards for the stream: 631
No cards for the stream: 643
No cards for the stream: 644
No cards for the stream: 645
No cards for the stream: 646
No cards for the stream: 647
No cards for the stream: 648
No cards for the stream: 649
No cards for the stream: 650
No cards for the stream: 651
No cards for the stream: 652
No cards for the stream: 653
No cards for the stream: 654
No cards for the stream: 655
No cards for the stream: 659
No cards for the stream: 682
No cards for the stream: 685
No cards for the stream: 701
No cards for the stream: 757
No cards for the stream: 793
No cards for the stream: 795
No cards for the stream: 796
No cards for the stream: 797
No cards for the stream: 798


STREAMID  579       580       581   582   583       584   585       586   587  \
USERID                                                                          
950       0.0  0.000000  0.000000  0.00  12.0  0.000000  0.00  0.000000   0.0   
958       0.0  0.263158  0.000000  0.00   0.0  0.000000  0.00  0.000000   0.0   
959       0.0  1.631579  1.263158  0.35   0.0  0.545455  0.00  1.684211   0.0   
961       0.0  0.736842  0.473684  0.00   0.0  1.454545  0.00  1.789474   0.0   
963       3.5  1.894737  1.105263  1.20   0.0  2.545455  5.25  5.315789  10.2   

STREAMID       588 ...   786  787  788  789  790  793  795  796  797  798  
USERID             ...                                                     
950       0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
958       1.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
959       0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
961       0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
963       3.142857 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 136 columns]

### test the above values

In [61]:
streamId = 587

if cards_per_stream.get(streamId):
    print(cards_per_stream[streamId])
else:
    print("No card for the stream with streamId {0} exists".format(str(streamId)))

5


In [62]:
content_views_by_user.to_csv("../data/generated/content_views_by_user_by_card.csv")

In [63]:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()

# deep copy a data frame
content_views_by_user_copy = content_views_by_user.copy(deep = True)

scaled_content_views = scaler.fit_transform(content_views_by_user_copy)

scaled_content_views_df = pd.DataFrame(scaled_content_views, columns=list(content_views_by_user), index = list(content_views_by_user.index))

print(scaled_content_views_df.head(5))

scaled_content_views_df.to_csv("../data/generated/content_views_by_user_scaled.csv")
print("File created...")

          579       580      581       582  583       584       585       586  \
950  0.000000  0.000000  0.00000  0.000000  1.0  0.000000  0.000000  0.000000   
958  0.000000  0.011547  0.00000  0.000000  0.0  0.000000  0.000000  0.000000   
959  0.000000  0.071594  0.04888  0.016018  0.0  0.008119  0.000000  0.004416   
961  0.000000  0.032333  0.01833  0.000000  0.0  0.021651  0.000000  0.004692   
963  0.031532  0.083141  0.04277  0.054920  0.0  0.037889  0.032407  0.013937   

          587       588 ...   786  787  788  789  790  793  795  796  797  798  
950  0.000000  0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
958  0.000000  0.008202 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
959  0.000000  0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
961  0.000000  0.000000 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
963  0.022952  0.025776 ...   0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 136 columns]
Fil